From 5c16d99886c71083fc522106e03af649df6c834b Mon Sep 17 00:00:00 2001
From: Samuel Verschelde <stormi@laposte.net>
Date: Wed, 16 Jan 2019 15:05:37 +0100
Subject: [PATCH] Allow migrate_send during RPU

It is needed to avoid downtime during RPU when VMs are running on local
storage and is already allowed for cross-pool storage migration between
mismatching XAPI versions, provided the destination has a higher version.

Related to https://github.com/xapi-project/xen-api/issues/3764

Signed-off-by: Samuel Verschelde <stormi@laposte.net>
---
 ocaml/tests/test_vm_check_operation_error.ml | 6 +++---
 ocaml/xapi/xapi_globs.ml                     | 1 +
 2 files changed, 4 insertions(+), 3 deletions(-)

diff --git a/ocaml/tests/test_vm_check_operation_error.ml b/ocaml/tests/test_vm_check_operation_error.ml
index 81ef115ae3..6fcbacd6e5 100644
--- a/ocaml/tests/test_vm_check_operation_error.ml
+++ b/ocaml/tests/test_vm_check_operation_error.ml
@@ -118,13 +118,13 @@ let test_migration_allowed_when_cbt_enabled_vdis_are_not_moved () =
         (Xapi_vm_lifecycle.check_operation_error ~__context ~ref:vM ~op:`migrate_send ~strict:true)
     )
 
-let test_sxm_disallowed_when_rum () =
+let test_sxm_allowed_when_rum () =
   with_test_vm (fun __context vm_ref ->
     let master = Test_common.make_host __context () in
     let pool = Test_common.make_pool ~__context ~master () in
     Db.Pool.add_to_other_config ~__context ~self:pool ~key:Xapi_globs.rolling_upgrade_in_progress ~value:"x";
     compare_errors
-      (Some(Api_errors.not_supported_during_upgrade, [ ]))
+      None
       (Xapi_vm_lifecycle.check_operation_error ~__context ~ref:vm_ref ~op:`migrate_send ~strict:false);
     Db.Pool.remove_from_other_config ~__context ~self:pool ~key:Xapi_globs.rolling_upgrade_in_progress;
     compare_errors
@@ -136,5 +136,5 @@ let test =
   [ "test_null_vdi", `Quick, test_null_vdi
   ; "test_operation_checks_allowed", `Quick, test_operation_checks_allowed
   ; "test_migration_allowed_when_cbt_enabled_vdis_are_not_moved", `Quick, test_migration_allowed_when_cbt_enabled_vdis_are_not_moved
-  ; "test_sxm_disallowed_when_rum", `Quick, test_sxm_disallowed_when_rum
+  ; "test_sxm_allowed_when_rum", `Quick, test_sxm_allowed_when_rum
   ]
diff --git a/ocaml/xapi/xapi_globs.ml b/ocaml/xapi/xapi_globs.ml
index 352da7c8d9..dab56a8d8c 100644
--- a/ocaml/xapi/xapi_globs.ml
+++ b/ocaml/xapi/xapi_globs.ml
@@ -377,6 +377,7 @@ let rpu_allowed_vm_operations = [
   `hard_shutdown;
   `import;
   `make_into_template;
+  `migrate_send;
   `pause;
   `pool_migrate;
   `power_state_reset;
